libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -280,13 +280,38 @@ bool ExpressionCompiler::visit(Assignment const& _assignment)
 	_assignment.rightHandSide().accept(*this);
 	// Perform some conversion already. This will convert storage types to memory and literals
 	// to their actual type, but will not convert e.g. memory to storage.
-	Type const* rightIntermediateType;
-	if (op != Token::Assign && TokenTraits::isShiftOp(binOp))
-		rightIntermediateType = _assignment.rightHandSide().annotation().type->mobileType();
-	else
-		rightIntermediateType = _assignment.rightHandSide().annotation().type->closestTemporaryType(
-			_assignment.leftHandSide().annotation().type
-		);
+	function<Type const*(Type const*, Type const*)> closestType = [op, binOp, &closestType](
```

Comment:
```suggestion
	auto closestType = [op, binOp, &closestType](
```

The type of a lambda is *not* ``std::function`` and ``std::function`` has serious overhead compared to the lambda-closure type itself.

Reply:
I created file level helper. It doesn't work with `auto` as it is recursive and needs type in order to be passed via capture list...

Reply:
There's a funny trick to define local recursive lambdas: you don't have them capture themselves, but you pass them to themselves as arguments :-)
```
auto fib = [](int _n, auto&& _recurse) { return _n < 2 ? _n : _recurse(_n - 1) + _recurse(_n - 2); }
int n = fib(1, fib);
```
or even:
```
auto fib = [](int _n) {
  auto fibImpl = [](int _n, auto&& _recurse) { return _n < 2 ? _n : _recurse(_n - 1) + _recurse(_n - 2); }
  return fibImpl(_n, fibImpl);
};
int n = fib(1);
```

But file-level sounds better here anyways :-).

Reply:
`s/funny/horrible/`

Reply:
I don't know what's not to like about it :-).

Reply:
Especially compared to what that innocent looking constructor of an ``std::function`` actually does, if one considers that as an alternative :-). That involves creating type-erased management objects with a v-table that has virtual destructor and copy functions and what-not... whereas the recursive lambda should actually result in the same code as a proper function.
But sure: most of the time a proper function is probably generally nicer :-).